pkg/3scale/amp/operator/backend_options_provider.go

```diff
@@ -0,0 +1,144 @@
+package operator
+
+import (
+	"fmt"
+
+	"github.com/3scale/3scale-operator/pkg/3scale/amp/component"
+	appsv1alpha1 "github.com/3scale/3scale-operator/pkg/apis/apps/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type OperatorBackendOptionsProvider struct {
+	apimanagerSpec *appsv1alpha1.APIManagerSpec
+	namespace      string
+	client         client.Client
+	backendOptions *component.BackendOptions
+	secretSource   *SecretSource
+}
+
+func NewOperatorBackendOptionsProvider(apimanagerSpec *appsv1alpha1.APIManagerSpec, namespace string, client client.Client) *OperatorBackendOptionsProvider {
+	return &OperatorBackendOptionsProvider{
+		apimanagerSpec: apimanagerSpec,
+		namespace:      namespace,
+		client:         client,
+		backendOptions: component.NewBackendOptions(),
+		secretSource:   NewSecretSource(client, namespace),
+	}
+}
+
+func (o *OperatorBackendOptionsProvider) GetBackendOptions() (*component.BackendOptions, error) {
+	o.backendOptions.AppLabel = *o.apimanagerSpec.AppLabel
+	o.backendOptions.TenantName = *o.apimanagerSpec.TenantName
+	o.backendOptions.WildcardDomain = o.apimanagerSpec.WildcardDomain
+
+	err := o.setSecretBasedOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	o.setResourceRequirementsOptions()
+	o.setReplicas()
+
+	err = o.backendOptions.Validate()
```

Comment:
* Might it happen that if we modify an existing "valid" BackendOptions the struct becomes an "invalid" struct from the validator point of view?

Reply:
Yes

With this implementation, options returned by the OptionProvider are enforced to be valid. What the receiver does with those options does not concern to the OptionProvider.

Reply:
If options received by the OptionProvider are modified for any reasons, you can always call `Validate()` method to check if they are still valid. `Validate()` method belongs to the options struct, not to the OptionProvider.

Reply:
:+1: